### PR TITLE
[sival/aes] Introduce AES Stall Status Bit test

### PR DESCRIPTION
**Pull request by github-gcontributor**
This commit introduces the AES stall test (chip_sw_stall_test) to
 verify the stall bit getting asserted when scenario of back pressure
is created.

- Intial Checkin: sw/device/tests/aes_stall_test.c
- Updated File: sw/device/tests/BUILD
- Updated File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson

### CODE REVIEW

**Comment by github-gcontributor** [comment]
Hi @moidx,
As discussed in the last meeting, I added stall test and created a new PR.
Kindly review.
Thanks.

File: sw/device/tests/aes_stall_test.c [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+
+    // Load data
+    // LOG_INFO("Loading plaintext block %d", i);
+    CHECK_DIF_OK(dif_aes_load_data(&aes, plain_text[i]));
+    num_of_blocks_loaded++;
+    LOG_INFO("Block %u Data loaded successfully", i);
+
+    // Check for STALL condition
+    CHECK_DIF_OK(dif_aes_get_status(&aes, kDifAesStatusStall, &stall));
+    if (stall) {
```
Comment (nasahlpa) [inline sw/device/tests/aes_stall_test.c:126]:
```suggestion
    TRY_CHECK(stall, "AES module not stalled as expected.");
```
As mentioned already in the other PR, you really should avoid just printing an error message if the expected behavior is not seen. Instead, you should abort the test and return an error. Then, the engineer executing the test immediately sees that the test failed.

File: sw/device/tests/aes_stall_test.c [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+      // Read output data
+      CHECK_DIF_OK(dif_aes_read_output(&aes, &cipher_text[num_of_blocks_read]));
+      num_of_blocks_read++;
+
+      // Check if STALL bit is cleared
+      CHECK_DIF_OK(dif_aes_get_status(&aes, kDifAesStatusStall, &stall));
+      if (!stall) {
+        LOG_INFO("Block %u Data read successfully", num_of_blocks_read - 1);
+      }
```
Comment (nasahlpa) [inline sw/device/tests/aes_stall_test.c:153]:
```suggestion
      TRY_CHECK(!stall, "Block not successfully read.");
```

File: sw/device/tests/aes_stall_test.c [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+  for (uint32_t i = 0; i < kAesNumBlocks; ++i) {
+    // Wait for input ready
+    input_ready = false;
+    do {
+      CHECK_DIF_OK(
+          dif_aes_get_status(&aes, kDifAesStatusInputReady, &input_ready));
+    } while (!input_ready);
+
+    // Load data
+    // LOG_INFO("Loading plaintext block %d", i);
```
Comment (nasahlpa) [inline sw/device/tests/aes_stall_test.c:119]:
```suggestion
```

**Review by nasahlpa — COMMENTED** [review]
Thanks for implementing this test. Mostly looks good to me.

File: sw/device/tests/aes_stall_test.c (hunk 1 above)
Comment (github-gcontributor) [reply sw/device/tests/aes_stall_test.c:126]: Noted! I need to change my coding style.

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson [hunk 4]
```diff
@@ -217,6 +217,51 @@
...
+            '''
+      features: [
+        "AES.MODE.ECB",
+        "AES.STATUS.STALL",
+      ]
+      stage: V2
+      si_stage: SV3
+      lc_states: ["PROD"]
+      tests: ["chip_sw_aes_stall_test"]
+      bazel: ["//sw/device/tests:aes_stall_test.c"]
```
Comment (nasahlpa) [inline hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:263]:
```suggestion
      bazel: ["//sw/device/tests:aes_stall_test"]
```
The suffix should not be in the bazel tag.

File: sw/device/tests/aes_stall_test.c (hunk 3 above)
Comment (nasahlpa) [reply sw/device/tests/aes_stall_test.c:119]: This is not resolved - the commented code still is in the repository.

File: sw/device/tests/aes_stall_test.c [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      .ctrl_aux_lock = false,
+  };
+
+  dif_aes_data_t plain_text[kAesNumBlocks];
+  dif_aes_data_t cipher_text[kAesNumBlocks];
+
+  // Initialize plaintext data dynamically
+  // Create plaintext with random data
+  dif_rv_core_ibex_t rv_core_ibex;
+  CHECK_DIF_OK(dif_rv_core_ibex_init(
```
Comment (engdoreis) [inline sw/device/tests/aes_stall_test.c:82]: You mix the macros `TRY` and `CHECK_DIF_OK`, could please use only one for consistency? The `TRY` macros is preferable.

**Review by engdoreis — COMMENTED** [review]
Overall LGTM, just a minor comment below:

File: sw/device/tests/aes_stall_test.c (hunk 5 above)
Comment (github-gcontributor) [reply sw/device/tests/aes_stall_test.c:82]: Noted!

File: hw/top_earlgrey/data/ip/chip_aes_testplan.hjson (hunk 4 above)
Comment (github-gcontributor) [reply hw/top_earlgrey/data/ip/chip_aes_testplan.hjson:263]: Noted.  I shall make this change in my repo and checkin.

File: sw/device/tests/aes_stall_test.c (hunk 3 above)
Comment (github-gcontributor) [reply sw/device/tests/aes_stall_test.c:119]: My bad. Somehow the change didn't go through.  I shall checkin with other review comments.

File: sw/device/tests/aes_stall_test.c (hunk 5 above)
Comment (nasahlpa) [reply sw/device/tests/aes_stall_test.c]: I think @engdoreis was also referring to other occurrences of `CHECK_DIF_OK` and not only the one in his comment.

Comment (github-gcontributor) [reply sw/device/tests/aes_stall_test.c]: Oh Okie.  
Just made the final updates and checkedin.
Hope this is the last thing before the approval :)

**Review by nasahlpa — APPROVED** [review]
Thanks, LGTM!

**Review by engdoreis — APPROVED** [review]
Thanks for the fixes, I re-triggered the filed CI jobs so we can merge it when it's green.

**Comment by github-gcontributor** [comment]
Hi @engdoreis,
Anything required from my side on this?

**Comment by engdoreis** [comment]
Hi @github-gcontributor, why did you close the PR? In case it was a mistake, can you please reopen it?

**Comment by github-gcontributor** [comment]
Hi @engdoreis,
I do not remember closing it.  I believe it might be a part of automation (auto-closure) after merge or incorrectly closed by GitHub.
Same case with aes_force_prng PR as well.

Let me see if I can reopen it.

**Comment by github-gcontributor** [comment]
I couldn't re-open the PR#25082.
Hence, created a new PR#25564
https://github.com/lowRISC/opentitan/pull/25564
Let me know if this works.

**Comment by nasahlpa** [comment]
Somehow this PR does not contain a commit? I think this is the reason why you cannot re-open it?
I'd honestly would prefer merging this PR such that we have the comment history. But if it does not work I can approve the new PR.

**Pull request closed without merging**